format-xls/src/main/java/io/cdap/plugin/format/xls/input/XlsInputFormat.java

```diff
@@ -0,0 +1,191 @@
+    // Map key that represents the row index.
+    private LongWritable key;
+    // Map value that represents an excel row
+    private StructuredRecord value;
+    private Sheet workSheet;
+    // InputStream handler for Excel files.
+    private FSDataInputStream fileIn;
+    // Specifies the row index.
+    private int rowIndex;
+    // Specifies last row num.
+    private int lastRowNum;
+    private Schema outputSchema;
+    private boolean terminateIfEmptyRow;
+    private boolean isRowNull;
+
+    @Override
+    public void initialize(InputSplit genericSplit, TaskAttemptContext context) throws IOException {
+
+      CombineFileSplit split = (CombineFileSplit) genericSplit;
+      Configuration jobConf = context.getConfiguration();
+      // Path of input file.
+      Path file = split.getPath(0);
+      String schema = context.getConfiguration().get("schema");
+      outputSchema = schema != null ? Schema.parseJson(schema) : null;
+      FileSystem fs = file.getFileSystem(jobConf);
+      fileIn = fs.open(split.getPath(0));
+
+      String sheet = jobConf.get(SHEET_NO);
+      String sheetValue = jobConf.get(SHEET_VALUE, "0");
+      terminateIfEmptyRow = jobConf.getBoolean(TERMINATE_IF_EMPTY_ROW, false);
+
+      try (Workbook workbook = WorkbookFactory.create(fileIn)) {
+        formulaEvaluator = workbook.getCreationHelper().createFormulaEvaluator();
+        formulaEvaluator.setIgnoreMissingWorkbooks(true);
+        formatter = new XlsInputFormatDataFormatter(formulaEvaluator);
+        // Check if user wants to access with name or number
+        if (sheet.equals(XlsInputFormatConfig.SHEET_NUMBER)) {
+          workSheet = workbook.getSheetAt(Integer.parseInt(sheetValue));
+        } else {
+          workSheet = workbook.getSheet(sheetValue);
+        }
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Exception while reading excel sheet. " + e.getMessage(), e);
+      }
+
+      lastRowNum = workSheet.getLastRowNum();
+      rowIndex = 0;
+      isRowNull = false;
+
+      boolean skipFirstRow = jobConf.getBoolean(NAME_SKIP_HEADER, true);
+      if (skipFirstRow) {
+        Preconditions.checkArgument(lastRowNum != -1, "No rows found on sheet %s", sheetValue);
+        rowIndex = 1;
+      }
+    }
+
+    @Override
+    public boolean nextKeyValue() {
+      // If any is true, then we stop processing.
+      if (rowIndex > lastRowNum || lastRowNum == -1 || (isRowNull && terminateIfEmptyRow)) {
+        return false;
+      }
+      // Get the next row.
+      Row row = workSheet.getRow(rowIndex);
+      key = new LongWritable(rowIndex);
+
+      StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+      List<Schema.Field> fields = outputSchema.getFields();
+
+      isRowNull = true;
+      for (int cellIndex = 0; cellIndex < row.getLastCellNum(); cellIndex++) {
+        if (cellIndex >= fields.size()) {
+         throw new IllegalArgumentException(
```

Comment:
Move this check outside the loop.

Reply:
How can i move `cellIndex` logic outside loop ?
`cellIndex` is part of this loop 🤔

Reply:
I mean instead of checking it inside the loop you can do it before the loop, something like:
```suggestion
if (row.getLastCellNum() > fields.size()) {
         throw new IllegalArgumentException(String.format("Schema contains less fields than the number of columns in the excel file. Schema fields: %s, Excel columns: %s", fields.size(), row.getLastCellNum()));
}
```